File: hs/67.hs
-- 7273
import Euler(readMatrix)

-- accumulate best path from the bottom-up
sumTri0 [[x]] = x
sumTri0 (x:y:ys) = sumTri0 (yy:ys)
    where xx = map (\(a,b) -> max a b) $ zip x (tail x)
          yy = map (\(a,b) -> a+b) $ zip xx y
sumTri0 _ = error "sumTri0: empty"
bestTriPath t = sumTri0 $ reverse t

main = do
    s <- readFile "../files/triangle.txt"
    putStrLn $ show $ bestTriPath $ readMatrix s

File: py/67.py
# 7273
tri = [map(int, line.split()) for line in file('../files/triangle.txt')][::-1]
for n in xrange(1, len(tri)):
    tri[n] = [max(tri[n-1][i:i+2]) + tri[n][i] for i in xrange(len(tri) - n)]
print tri[-1][0]
